src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -9703,6 +9659,66 @@ static bool isCandidateUnique(ref MethodSymbol? method, MethodSymbol candidate)
                 method = null;
                 return false;
             }
+
+            MethodSymbol? selectMethodForSignature(BoundMethodGroup node)
+            {
+                MethodSymbol? method = null;
+                if (node.ResultKind == LookupResultKind.Viable)
```

Comment:
:memo: the reason for this change is the following scenario which currently gives an error but should not: 
```
System.Action x = new C().M<int, int>;
x();

var z = new C().M<int, int>; // Error CS8917 The delegate type could not be inferred.

z();

public class C
{
    public void M<T>(C c) { }
}

public static class E
{
    public static void M<T, U>(this C c)
    {
        System.Console.Write("E.M<T, U> ");
    }
}
```

See the new behavior (no error) in test `MethodGroup_ScopeByScope_TypeArgumentsDoNotMatchInstanceMethod`.
The root for this behavior is this logic in [BindInstanceMemberAccess](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs#L7190-L7198) which stores all the instance methods we found (even though they have the wrong arity) into the BoundMethodGroup.